provider/apple.go

```diff
@@ -0,0 +1,529 @@
+		return
+	}
+	rest.RenderJSON(w, &u)
+
+}
+
+// LogoutHandler - GET /logout
+func (ah AppleHandler) LogoutHandler(w http.ResponseWriter, r *http.Request) {
+	if _, _, err := ah.JwtService.Get(r); err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusForbidden, err, "logout not allowed")
+		return
+	}
+	ah.JwtService.Reset(w)
+}
+
+// exchange sends the validation token request and gets access token and user claims
+// (e.g. https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens)
+func (ah *AppleHandler) exchange(ctx context.Context, code, redirectURI string, result *AppleVerificationResponse) error {
+
+	// check client_secret for valid and recreate new (client_secret JWT) if required
+	if tkn, err := jwt.Parse(ah.conf.clientSecret, ah.tokenKeyFunc); err != nil || tkn == nil {
+		ah.conf.clientSecret, err = ah.createClientSecret()
+		if err != nil {
+			return errors.Wrap(err, "client secret create failed")
+		}
+	}
+
+	data := url.Values{}
+	data.Set("client_id", ah.conf.ClientID)
+	data.Set("client_secret", ah.conf.clientSecret) // JWT signed with Apple private key
+	data.Set("code", code)
+	data.Set("redirect_uri", redirectURI) // redirect URL can't refer to localhost and must have trusted certificate and https protocol
+	data.Set("grant_type", "authorization_code")
+
+	client := http.Client{Timeout: time.Second * 5}
+	req, err := http.NewRequestWithContext(ctx, "POST", ah.endpoint.TokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("content-type", appleRequestContentType)
+	req.Header.Add("accept", AcceptJSONHeader)
+	req.Header.Add("user-agent", defaultUserAgent) // apple requires a user agent
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Trying to decode (unmarshal json) data of response
+	err = json.NewDecoder(res.Body).Decode(result)
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling data from apple service response failed")
+	}
+
+	// If above operation done successfully checking a response code and error descriptions, if one exist.
+	// Apple service will response either 200 (OK) or 400 (any error).
+	if res.StatusCode != http.StatusOK || result.Error != "" {
+		return fmt.Errorf("apple token service error: %s", result.Error)
+	}
+
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+
+	return err
+}
+
+// createClientSecret use for create the JWT client secret required to make requests to the Apple validation server.
+// for more details go to link: https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+func (ah *AppleHandler) createClientSecret() (string, error) {
+	if ah.conf.privateKey == nil {
+		return "", errors.New("private key can't be empty")
+	}
+	// Create the Claims
+	now := time.Now()
+	exp := now.Add(time.Second * 1800).Unix() // default value
```

Comment:
maybe more human-friendy form will be more readbale? lile time.Minute*30

Reply:
I define in seconds, because assume that a user want to define lifitime in seconds. Are you think remove lifitime redefine from the `AppleProvider` config too?

Reply:
this is just a duration, right? so, 30 minutes seems easier to read. 

> Are you think remove lifitime redefine from the AppleProvider config too?

I think this is a good idea.

Reply:
Ок.  I will remove it.